src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -1820,6 +1907,253 @@ public string DeleteCounter(DocumentsOperationContext context, string documentId
+                using (Slice.From(context.Allocator, changeVector, out var cv))
+                using (DocumentIdWorker.GetStringPreserveCase(context, collectionName.Name, out Slice collectionSlice))
+                {
+                    tvb.Add(counterTombstoneKeySlice);
+                    tvb.Add(Bits.SwapBytes(etag));
+                    tvb.Add(cv);
+                    tvb.Add(collectionSlice);
+                    tvb.Add(context.GetTransactionMarker());
+
+                    table.Set(tvb);
+                }
+            }
+        }
+
+        private static string ExtractCounterTombstoneChangeVector(DocumentsOperationContext context, ref TableValueReader reader)
+        {
+            var changeVectorPtr = reader.Read((int)CounterTombstonesTable.ChangeVector, out int changeVectorSize);
+            return Encoding.UTF8.GetString(changeVectorPtr, changeVectorSize);
+        }
+
+        public IEnumerable<CounterTombstoneDetail> GetCounterTombstonesForDoc(DocumentsOperationContext context, string docId)
+        {
+            var table = new Table(CounterTombstonesSchema, context.Transaction.InnerTransaction);
+            using var dispose = DocumentIdWorker.GetSliceFromId(context, docId, out var documentKeyPrefix, SpecialChars.RecordSeparator);
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach ((_, Table.TableValueHolder tvh) in table.SeekByPrimaryKeyPrefix(documentKeyPrefix, Slices.Empty, 0))
+            {
+                var item = TableValueToCounterTombstoneDetail(context, ref tvh.Reader);
+                yield return item;
+            }
+        }
+
+        public IEnumerable<CounterTombstoneDetail> GetCounterTombstonesFrom(DocumentsOperationContext context, long etag, long toEtag = long.MaxValue)
+        {
+            var table = new Table(CounterTombstonesSchema, context.Transaction.InnerTransaction);
+
+            // ReSharper disable once LoopCanBeConvertedToQuery
+            foreach (var result in table.SeekForwardFrom(CounterTombstonesSchema.FixedSizeIndexes[AllCounterTombstonesEtagSlice], etag, 0))
+            {
+                var item = TableValueToCounterTombstoneDetail(context, ref result.Reader);
+                if (item.Etag > toEtag)
+                    yield break;
+                yield return item;
+            }
+        }
+
+        public long PurgeCountersAndCounterTombstones(DocumentsOperationContext context, string collection, long upto, long numberOfEntriesToDelete)
+        {
+            var collectionName = _documentsStorage.GetCollection(collection, throwIfDoesNotExist: false);
+            if (collectionName == null)
+                return 0;
+
+            var deletedCounters = PurgeCounters(upto, context, collectionName, numberOfEntriesToDelete);
+            var deletedTombstoneEntries = PurgeCounterTombstones(upto, context, collectionName, numberOfEntriesToDelete - deletedCounters);
```

Comment:
what if you didn't delete everything in a single batch? 
The result of `numberOfEntriesToDelete - deletedCounters` might be zero, is that okay?

Reply:
I think that's fine. If there is something left to delete -> we get to it on the next batches. 
For example: if _batch size = 1024_ & _2024 counter tombstone entries_ -> in the first batch - delete 1024 from the counters table (`numberOfEntriesToDelete - deletedCounters = 0`), but the `min Etag` for counter tombstones hasn't changed so in the next batch we remove the next 1000 counters and 24 entries from the counter tombstones table. Then, in the last batch, we'll remove all remaining entries from counter tombstones table (no counters to delete this time) and stop.

Reply:
okay, but let's have a test to verify

Reply:
done